File: clients/sdk-cookbook/src/recipes/approve-or-deny.ts
/**
 * Recipe: approve or deny the agent's permission request.
 *
 * Runs against `muse-conformance serve-fixture` playing two committed golden
 * transcripts — `schema/msp/transcripts/approval-round-trip` (allow) and
 * `approval-deny-round-trip` (deny) — so this recipe is deterministic and
 * needs no credentials and no model. It plays BOTH arms because the deny path
 * is the one an integrator gets wrong, and the quickstart never shows it.
 *
 * What it teaches (and what the docs page walks through):
 *  - the host asks TWICE, for two different audiences: `approval/requested` is
 *    a notification for your UI, and `approval/request` is a JSON-RPC request
 *    the SDK must answer. If you never register a server-request handler, the
 *    SDK answers "method not found" on your behalf — so answer it.
 *  - answer with `approval/decide`, picking a `choiceId` the host OFFERED in
 *    `availableChoices`. Do not invent one, and do not hardcode a choice the
 *    host may not offer for this subject.
 *  - the ack is not the outcome. `approval/resolved` is where the decision,
 *    the policy result, and any session-scoped rule amendment land.
 *  - denying is a normal answer, not an error: the turn keeps running and the
 *    agent tells the user what it did not do.
 *
 * Fixture mechanics, so the code below reads honestly: serve-fixture matches
 * our frames structurally against the transcript's client lines and then
 * replays the recorded server lines byte-exactly — except that a reply to
 * one of our requests comes back under the id WE sent (#25143), so the SDK's
 * own request-id minting correlates replies here exactly as it does against
 * a real host. Acks still echo the RECORDED commandId — and
 * `Connection.command` verifies that echo — so each command below passes the
 * transcript's commandId explicitly. Values we learn from the wire
 * (sessionId, approvalId, requirementId) are used as received, exactly as a
 * real client uses them.
 */

import { Host, arrayAt, equals, objectAt, requireHost, stringAt, within } from "../kit/host.js";
import { runJourney } from "../kit/segments.js";
import type { JourneyReport, Segment } from "../kit/segments.js";
import type { Recipe, RecipeHosts } from "../runner.js";

const HANDSHAKE_BUDGET_MS = 30_000;
const COMMAND_BUDGET_MS = 30_000;
const CLOSE_BUDGET_MS = 30_000;

/** The transcripts' own client-side values; see the module comment. */
const WORKSPACE_ROOT = "/home/me/src/proj";
const PROMPT = "Update Cargo.toml to bump the version";
const SESSION_START_COMMAND_ID = "0198f0ab-9999-7000-8000-0000000000c1";
const TURN_START_COMMAND_ID = "018f6a1e-9b3c-7c21-a54a-2f30bd3c9f10";
const APPROVAL_DECIDE_COMMAND_ID = "018f6a2a-3333-7abc-8def-00000000d001";

/** One arm of the recipe: the same round trip, a different answer. */
interface Arm {
  /** Segment-id prefix, so the report names which arm a line belongs to. */
  readonly id: string;
  /** The transcript directory name under the corpus root. */
  readonly scenario: string;
  /** The choice this arm picks; asserted to be one the host offered. */
  readonly choiceId: string;
  /** Sent only when the chosen option accepts feedback. */
  readonly feedback?: string;
  /** The decision `approval/resolved` must report. */
  readonly decision: string;
  /** The policy result `approval/resolved` must report. */
  readonly policyResult: string;
}

const ALLOW: Arm = {
  id: "allow",
  scenario: "approval-round-trip",
  choiceId: "allow_session",
  decision: "approvedForSession",
  policyResult: "allow",
};

const DENY: Arm = {
  id: "deny",
  scenario: "approval-deny-round-trip",
  choiceId: "abort",
  feedback: "Do not modify the manifest",
  decision: "abort",
  policyResult: "deny",
};

interface Context {
  readonly conformanceBin: string;
  readonly transcriptRoot: string;
  host?: Host;
  sessionId?: string;
  turnId?: string;
  approvalId?: string;
  requirementId?: unknown;
  /** Settles once the reply to `approval/request` has been written. */
  answered?: Promise<void>;
}

function requireString(value: string | undefined, what: string): string {
  if (value === undefined) throw new Error(`${what} is not known: an earlier segment did not finish`);
  return value;
}

function armSegments(arm: Arm): ReadonlyArray<Segment<Context>> {
  return [
    {
      id: `${arm.id}-spawn`,
      title: `Spawn the canned host playing the ${arm.scenario} transcript`,
      async run(context) {
        // Reclaim before overwriting the slot. A previous arm whose drain
        // threw leaves its child in `context.host`; assigning over it would
        // orphan that child with no warning and leave later segments of a
        // failed arm talking to the stale host (PR #25153 review).
        await context.host?.abandon(CLOSE_BUDGET_MS);
        context.host = undefined;
        // THE authoritative reset of everything an arm learns from its own
        // wire. It lives here, not in `-drain`, because `-drain` runs after
        // close assertions that can throw — a failed drain would then leave
        // this arm's state for the next one, and a segment would burn its
        // whole budget on a stale id instead of failing immediately with
        // "an earlier segment did not finish". Keeping the only copy here
        // also means the two lists cannot drift apart, which they already
        // had (PR #25153 review).
        context.sessionId = undefined;
        context.turnId = undefined;
        context.approvalId = undefined;
        context.requirementId = undefined;
        context.answered = undefined;
        const host = await Host.spawn(
          {
            command: context.conformanceBin,
            args: ["serve-fixture", "--transcript", `${context.transcriptRoot}/${arm.scenario}`],
            env: { PATH: process.env["PATH"] ?? "/usr/bin:/bin" },
            clientInfo: { name: "conformance", version: "0.0.0" },
          },
          HANDSHAKE_BUDGET_MS,
        );
        // Answer `approval/request` — the JSON-RPC request half of the ask.
        // The reply is an empty acknowledgement: it means "this client is
        // handling the approval", NOT "approved". The actual answer travels
        // separately as `approval/decide`. Registered before any turn starts,
        // because the host may ask the moment the agent reaches for a tool.
        //
        // The two writes are ORDERED, and the fixture enforces it: the
        // transcript records the reply before the decide, so a decide that
        // overtakes the reply is a frame divergence at that line, not a
        // tolerated reordering. The connection writes the reply on the
        // microtask AFTER this handler settles, so the signal below is
        // deliberately scheduled one hop late — resolving it inline would let
        // the journey continue while the reply write was still unqueued. The
        // decide segment then flushes, so the reply is on its way to the host
        // before the decision is even written.
        let replied = (): void => {};
        context.answered = new Promise<void>((resolve) => {
          replied = () => {
            resolve();
          };
        });
        host.msp.connection.onServerRequest(async (request) => {
          // Check the method, exactly as the page teaches: one handler serves
          // the whole connection, so a bare `{}` ack would also tell the host
          // this client is handling requests it has no UI for. The canned
          // transcripts only ever send `approval/request`, so the throw arm
          // is never taken during replay (PR #25153 review).
          if (request.method !== "approval/request") {
            throw new Error(`unhandled server request: ${request.method}`);
          }
          queueMicrotask(() => {
            replied();
          });
          return {};
        });
        context.host = host;
      },
    },
    {
      id: `${arm.id}-turn`,
      title: "Start a session and a turn that will need permission",
      async run(context) {
        const host = requireHost(context);
        const started = await host.msp.connection.command(
          "session/start",
          { workspaceRoot: WORKSPACE_ROOT },
          { maxAttempts: 1, commandId: SESSION_START_COMMAND_ID },
        );
        const session = objectAt(started, "session", "session/start result");
        const sessionId = stringAt(session, "sessionId", "session/start session");
        const ack = await host.msp.connection.command(
          "turn/start",
          { sessionId, input: [{ type: "text", text: PROMPT }] },
          { maxAttempts: 1, commandId: TURN_START_COMMAND_ID },
        );
        equals(ack["status"], "accepted", "turn/start ack status");
        context.sessionId = sessionId;
        context.turnId = stringAt(ack, "turnId", "turn/start ack");
      },
    },
    {
      id: `${arm.id}-request`,
      title: "Take the request off the wire and read the choices the host offers",
      async run(context) {
        const host = requireHost(context);
        const turnId = requireString(context.turnId, "the turn id");
        // Race the turn's own terminal: without it, a turn that ended WITHOUT
        // ever asking for permission would sit here until the budget expired
        // and report a timeout instead of what actually happened.
        const event = await host.waitFor(
          "an approval/requested notification",
          COMMAND_BUDGET_MS,
          (notification) =>
            notification.method === "approval/requested" ||
            (notification.method === "turn/completed" && notification.params["turnId"] === turnId),
        );
        if (event.method !== "approval/requested") {
          throw new Error(
            `the turn ended (terminal ${JSON.stringify(event.params["terminal"])}) without asking for approval`,
          );
        }
        equals(event.params["turnId"], turnId, "the approval's turn id");

        // What a UI puts in front of the user: which tool, acting on what.
        const subject = objectAt(event.params, "subject", "approval/requested params");
        stringAt(subject, "kind", "approval/requested subject");
        stringAt(event.params, "toolName", "approval/requested params");

        // Pick from what the host OFFERED. A hardcoded choiceId is the bug
        // this assertion exists to catch: the offered set depends on the
        // subject, so a client that assumes one can send an answer the host
        // will reject.
        const choices = arrayAt(event.params, "availableChoices", "approval/requested params");
        const offered = choices.map((choice) =>
          stringAt(choice as Record<string, unknown>, "choiceId", "an availableChoices entry"),
        );
        if (!offered.includes(arm.choiceId)) {
          throw new Error(
            `the host did not offer "${arm.choiceId}"; it offered ${JSON.stringify(offered)}`,
          );
        }
        if (arm.feedback !== undefined) {
          // Feedback rides along only where the host says it is accepted.
          const chosen = choices.find(
            (choice) => (choice as Record<string, unknown>)["choiceId"] === arm.choiceId,
          ) as Record<string, unknown>;
          equals(chosen["acceptsFeedback"], true, `whether "${arm.choiceId}" accepts feedback`);
        }

        context.approvalId = stringAt(event.params, "approvalId", "approval/requested params");
        // The requirement being answered, quoted back verbatim on the decide.
        context.requirementId = objectAt(
          event.params,
          "currentRequirementId",
          "approval/requested params",
        );
      },
    },
    {
      id: `${arm.id}-decide`,
      title: `Answer with "${arm.choiceId}"`,
      async run(context) {
        const host = requireHost(context);
        const sessionId = requireString(context.sessionId, "the session id");
        const approvalId = requireString(context.approvalId, "the approval id");
        // Reply first, decision second — see the spawn segment. Bounded, so a
        // handler that never fires reports that instead of hanging.
        if (context.answered === undefined) throw new Error("`spawn` did not finish");
        await within("the reply to approval/request", COMMAND_BUDGET_MS, context.answered);
        await host.msp.connection.flush();
        const ack = await host.msp.connection.command(
          "approval/decide",
          {
            sessionId,
            approvalId,
            requirementId: context.requirementId,
            choiceId: arm.choiceId,
            ...(arm.feedback === undefined ? {} : { feedback: arm.feedback }),
          },
          { maxAttempts: 1, commandId: APPROVAL_DECIDE_COMMAND_ID },
        );
        equals(ack["status"], "accepted", "approval/decide ack status");
        equals(ack["approvalId"], approvalId, "the approval id the ack echoes");
        // `terminal: true` means this approval will not come back for more:
        // the requirement is fully answered.
        equals(ack["terminal"], true, "whether the decision settles the approval");
      },
    },
    {
      id: `${arm.id}-resolved`,
      title: "Read the outcome off approval/resolved, then let the turn finish",
      async run(context) {
        const host = requireHost(context);
        const approvalId = requireString(context.approvalId, "the approval id");
        const turnId = requireString(context.turnId, "the turn id");
        const resolved = await host.waitFor(
          "the approval/resolved notification",
          COMMAND_BUDGET_MS,
          (notification) =>
            notification.method === "approval/resolved" &&
            notification.params["approvalId"] === approvalId,
        );
        // The ack said "accepted". THIS says what was decided.
        equals(resolved.params["decision"], arm.decision, "the decision");
        equals(resolved.params["policyResult"], arm.policyResult, "the policy result");
        equals(resolved.params["resolvedBy"], "user", "who resolved the approval");

        // A session-scoped choice also amends policy, so the same subject is
        // not asked about again for the rest of the session. A once-scoped or
        // denying choice carries no amendment — that difference IS the scope.
        const amendment = resolved.params["amendment"];
        if (arm.choiceId === "allow_session") {
          const durability = objectAt(resolved.params, "amendment", "approval/resolved params");
          equals(durability["durability"], "session", "the amendment's durability");
        } else if (amendment !== undefined) {
          throw new Error(
            `a "${arm.choiceId}" decision should amend no policy, but carried ${JSON.stringify(amendment)}`,
          );
        }

        // Denied or allowed, the turn runs on to its own terminal. A denial is
        // an answer, not a failure.
        const completed = await host.waitFor(
          "the turn/completed notification",
          COMMAND_BUDGET_MS,
          (notification) =>
            notification.method === "turn/completed" && notification.params["turnId"] === turnId,
        );
        equals(completed.params["terminal"], "completed", "the turn's terminal");

        if (arm.policyResult === "deny") {
          // ...and the user is told. A denied tool call that ends in silence
          // is a UI that looks broken.
          const spoke = host.notifications().some((notification) => {
            if (notification.method !== "item/completed") return false;
            const item = notification.params["item"] as Record<string, unknown> | undefined;
            return (
              item?.["kind"] === "agentMessage" &&
              typeof item["text"] === "string" &&
              item["text"].length > 0
            );
          });
          if (!spoke) {
            throw new Error("the denied turn ended without the agent saying what it did not do");
          }
        }
      },
    },
    {
      id: `${arm.id}-drain`,
      title: "Close stdin and let the fixture host exit cleanly",
      async run(context) {
        const host = requireHost(context);
        const exit = await host.close(CLOSE_BUDGET_MS);
        equals(exit.code, 0, "the fixture host's exit code after stdin EOF");
        // Only the host slot, so journey-end teardown does not re-abandon a
        // host this segment already closed cleanly. Everything this arm
        // learned from its wire is reset by the NEXT arm's spawn, which is
        // the one place that runs whether or not this drain succeeded.
        context.host = undefined;
      },
    },
  ];
}

const SEGMENTS: ReadonlyArray<Segment<Context>> = [...armSegments(ALLOW), ...armSegments(DENY)];

export const approveOrDeny: Recipe = {
  id: "approve-or-deny",
  title: "Approve or deny the agent's permission request",
  docsPage: "developer-docs/src/content/docs/cookbook/approve-or-deny-a-tool-call.mdx",
  needs: ["conformanceBin"],
  async run(hosts: RecipeHosts): Promise<JourneyReport> {
    const conformanceBin = hosts.conformanceBin;
    if (conformanceBin === undefined) throw new Error("conformanceBin is required");
    const context: Context = {
      conformanceBin,
      transcriptRoot: hosts.transcriptRoot,
    };
    return await runJourney(SEGMENTS, context, async (owned) => {
      await owned.host?.abandon(CLOSE_BUDGET_MS);
    });
  },
};

File: python/clients/sdk-cookbook-py/src/cookbook_recipes/recipes/approve_or_deny.py
"""Recipe twin: approve or deny the agent's permission request.

Runs against ``muse-conformance serve-fixture`` playing two committed golden
transcripts — ``approval-round-trip`` (allow) and ``approval-deny-round-trip``
(deny) — so it is deterministic and needs no credentials and no model. It
plays BOTH arms because the deny path is the one an integrator gets wrong,
and the quickstart never shows it.

What it teaches (the docs page walks the same points): the host asks TWICE,
for two audiences (``approval/requested`` for your UI, ``approval/request``
as a JSON-RPC request the client must answer — reply first, decide second);
answer with a ``choiceId`` the host OFFERED; the ack is not the outcome —
``approval/resolved`` is; and denying is a normal answer, not an error.
"""

from __future__ import annotations

import asyncio
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Mapping

from ..kit import (
    array_at,
    equals,
    Host,
    JourneyReport,
    object_at,
    require_host,
    run_journey,
    Segment,
    spawn_fixture_host,
    string_at,
    within,
)
from ..runner import Recipe, RecipeHosts

HANDSHAKE_BUDGET_MS = 30_000
COMMAND_BUDGET_MS = 30_000
CLOSE_BUDGET_MS = 30_000

# The transcripts' own client-side values; see the module docstring.
WORKSPACE_ROOT = "/home/me/src/proj"
PROMPT = "Update Cargo.toml to bump the version"
SESSION_START_COMMAND_ID = "0198f0ab-9999-7000-8000-0000000000c1"
TURN_START_COMMAND_ID = "018f6a1e-9b3c-7c21-a54a-2f30bd3c9f10"
APPROVAL_DECIDE_COMMAND_ID = "018f6a2a-3333-7abc-8def-00000000d001"


@dataclass(frozen=True)
class Arm:
    """One arm of the recipe: the same round trip, a different answer.

    Attributes:
        id: Segment-id prefix, so the report names which arm a line is.
        scenario: The transcript directory name under the corpus root.
        choice_id: The choice this arm picks; asserted to be one the host
            offered.
        decision: The decision ``approval/resolved`` must report.
        policy_result: The policy result ``approval/resolved`` must report.
        feedback: Sent only when the chosen option accepts feedback.
    """

    id: str
    scenario: str
    choice_id: str
    decision: str
    policy_result: str
    feedback: str | None = None


ALLOW = Arm(
    id="allow",
    scenario="approval-round-trip",
    choice_id="allow_session",
    decision="approvedForSession",
    policy_result="allow",
)

DENY = Arm(
    id="deny",
    scenario="approval-deny-round-trip",
    choice_id="abort",
    feedback="Do not modify the manifest",
    decision="abort",
    policy_result="deny",
)


@dataclass
class Context:
    """The recipe's own state, reset by each arm's spawn segment."""

    conformance_bin: str
    transcript_root: str
    host: Host | None = None
    session_id: str | None = None
    turn_id: str | None = None
    approval_id: str | None = None
    requirement_id: Mapping[str, Any] | None = None
    answered: "asyncio.Event | None" = None


def _known(value: str | None, what: str) -> str:
    if value is None:
        raise RuntimeError(f"{what} is not known: an earlier segment did not finish")
    return value


def _arm_segments(arm: Arm) -> tuple[Segment[Context], ...]:
    async def spawn(context: Context) -> None:
        # Reclaim before overwriting the slot: a previous arm whose drain
        # threw leaves its child in context.host, and assigning over it would
        # orphan that child and leave later segments talking to a stale host.
        if context.host is not None:
            await context.host.abandon(CLOSE_BUDGET_MS)
            context.host = None
        # THE authoritative reset of everything an arm learns from its own
        # wire — here, not in drain, because drain runs after assertions that
        # can throw and would then leave stale state for the next arm.
        context.session_id = None
        context.turn_id = None
        context.approval_id = None
        context.requirement_id = None
        context.answered = None
        host = await spawn_fixture_host(
            conformance_bin=context.conformance_bin,
            transcript_dir=Path(context.transcript_root) / arm.scenario,
            env={"PATH": os.environ.get("PATH", "/usr/bin:/bin")},
            client_info={"name": "conformance", "version": "0.0.0"},
            budget_ms=HANDSHAKE_BUDGET_MS,
        )
        # Answer approval/request — the JSON-RPC request half of the ask. The
        # reply is an empty acknowledgement: "this client is handling the
        # approval", NOT "approved"; the answer travels separately as
        # approval/decide. Registered before any turn starts. The connection
        # enqueues the reply write synchronously after this handler returns
        # (a direct await, no loop hop), so setting the event inline is
        # enough: by the time the decide segment resumes, the reply is ahead
        # of the decide in the write tail, and the flush there orders both.
        answered = asyncio.Event()
        context.answered = answered

        async def handle(request: Mapping[str, Any]) -> Mapping[str, Any]:
            # Check the method, exactly as the page teaches: one handler
            # serves the whole connection. The canned transcripts only ever
            # send approval/request, so the raise arm never fires in replay.
            if request.get("method") != "approval/request":
                raise RuntimeError(f"unhandled server request: {request.get('method')}")
            answered.set()
            return {}

        host.connection.on_server_request(handle)
        context.host = host

    async def turn(context: Context) -> None:
        host = require_host(context.host, f"{arm.id}-spawn")
        started = await within("the session/start ack", COMMAND_BUDGET_MS, host.connection.command(
            "session/start",
            {"workspaceRoot": WORKSPACE_ROOT},
            max_attempts=1,
            command_id=SESSION_START_COMMAND_ID,
        ))
        session = object_at(started, "session", "session/start result")
        session_id = string_at(session, "sessionId", "session/start session")
        # The recorded client discovers the model catalog right at session
        # acquisition — one read-only model/list.
        await within("the model/list reply", COMMAND_BUDGET_MS, host.connection.request("model/list", {"sessionId": session_id}))
        ack = await within("the turn/start ack", COMMAND_BUDGET_MS, host.connection.command(
            "turn/start",
            {"sessionId": session_id, "input": [{"type": "text", "text": PROMPT}]},
            max_attempts=1,
            command_id=TURN_START_COMMAND_ID,
        ))
        equals(ack.get("status"), "accepted", "turn/start ack status")
        context.session_id = session_id
        context.turn_id = string_at(ack, "turnId", "turn/start ack")

    async def request(context: Context) -> None:
        host = require_host(context.host, f"{arm.id}-spawn")
        turn_id = _known(context.turn_id, "the turn id")
        # Race the turn's own terminal: without it, a turn that ended WITHOUT
        # asking for permission would burn the whole budget and report a
        # timeout instead of what actually happened.
        event = await host.wait_for(
            "an approval/requested notification",
            COMMAND_BUDGET_MS,
            lambda n: n.method == "approval/requested"
            or (n.method == "turn/completed" and n.params.get("turnId") == turn_id),
        )
        if event.method != "approval/requested":
            raise AssertionError(
                f"the turn ended (terminal {event.params.get('terminal')!r}) "
                "without asking for approval"
            )
        equals(event.params.get("turnId"), turn_id, "the approval's turn id")

        # What a UI puts in front of the user: which tool, acting on what.
        subject = object_at(event.params, "subject", "approval/requested params")
        string_at(subject, "kind", "approval/requested subject")
        string_at(event.params, "toolName", "approval/requested params")

        # Pick from what the host OFFERED. A hardcoded choiceId is the bug
        # this assertion exists to catch.
        choices = array_at(event.params, "availableChoices", "approval/requested params")
        offered = [
            string_at(choice, "choiceId", "an availableChoices entry") for choice in choices
        ]
        if arm.choice_id not in offered:
            raise AssertionError(
                f'the host did not offer "{arm.choice_id}"; it offered {offered!r}'
            )
        if arm.feedback is not None:
            # Feedback rides along only where the host says it is accepted.
            chosen = next(c for c in choices if c.get("choiceId") == arm.choice_id)
            equals(
                chosen.get("acceptsFeedback"),
                True,
                f'whether "{arm.choice_id}" accepts feedback',
            )

        context.approval_id = string_at(event.params, "approvalId", "approval/requested params")
        # The requirement being answered, quoted back verbatim on the decide.
        context.requirement_id = object_at(
            event.params, "currentRequirementId", "approval/requested params"
        )

    async def decide(context: Context) -> None:
        host = require_host(context.host, f"{arm.id}-spawn")
        session_id = _known(context.session_id, "the session id")
        approval_id = _known(context.approval_id, "the approval id")
        # Reply first, decision second — see the spawn segment. Bounded, so a
        # handler that never fires reports that instead of hanging.
        if context.answered is None:
            raise RuntimeError(f"`{arm.id}-spawn` did not finish")
        await within(
            "the reply to approval/request", COMMAND_BUDGET_MS, context.answered.wait()
        )
        await within("the reply flush", COMMAND_BUDGET_MS, host.connection.flush())
        params: dict[str, Any] = {
            "sessionId": session_id,
            "approvalId": approval_id,
            "requirementId": context.requirement_id,
            "choiceId": arm.choice_id,
        }
        if arm.feedback is not None:
            params["feedback"] = arm.feedback
        ack = await within("the approval/decide ack", COMMAND_BUDGET_MS, host.connection.command(
            "approval/decide", params, max_attempts=1, command_id=APPROVAL_DECIDE_COMMAND_ID
        ))
        equals(ack.get("status"), "accepted", "approval/decide ack status")
        equals(ack.get("approvalId"), approval_id, "the approval id the ack echoes")
        # terminal: true means this approval will not come back for more.
        equals(ack.get("terminal"), True, "whether the decision settles the approval")

    async def resolved(context: Context) -> None:
        host = require_host(context.host, f"{arm.id}-spawn")
        approval_id = _known(context.approval_id, "the approval id")
        turn_id = _known(context.turn_id, "the turn id")
        outcome = await host.wait_for(
            "the approval/resolved notification",
            COMMAND_BUDGET_MS,
            lambda n: n.method == "approval/resolved"
            and n.params.get("approvalId") == approval_id,
        )
        # The ack said "accepted". THIS says what was decided.
        equals(outcome.params.get("decision"), arm.decision, "the decision")
        equals(outcome.params.get("policyResult"), arm.policy_result, "the policy result")
        equals(outcome.params.get("resolvedBy"), "user", "who resolved the approval")

        # A session-scoped choice also amends policy; a once-scoped or denying
        # choice carries no amendment — that difference IS the scope.
        amendment = outcome.params.get("amendment")
        if arm.choice_id == "allow_session":
            durability = object_at(outcome.params, "amendment", "approval/resolved params")
            equals(durability.get("durability"), "session", "the amendment's durability")
        elif amendment is not None:
            raise AssertionError(
                f'a "{arm.choice_id}" decision should amend no policy, '
                f"but carried {amendment!r}"
            )

        # Denied or allowed, the turn runs on to its own terminal. A denial is
        # an answer, not a failure.
        completed = await host.wait_for(
            "the turn/completed notification",
            COMMAND_BUDGET_MS,
            lambda n: n.method == "turn/completed" and n.params.get("turnId") == turn_id,
        )
        equals(completed.params.get("terminal"), "completed", "the turn's terminal")

        if arm.policy_result == "deny":
            # ...and the user is told. A denied tool call that ends in silence
            # is a UI that looks broken.
            def _spoke(notification: Any) -> bool:
                if notification.method != "item/completed":
                    return False
                item = notification.params.get("item")
                return (
                    isinstance(item, dict)
                    and item.get("kind") == "agentMessage"
                    and isinstance(item.get("text"), str)
                    and len(item["text"]) > 0
                )

            if not any(_spoke(n) for n in host.notifications()):
                raise AssertionError(
                    "the denied turn ended without the agent saying what it did not do"
                )

    async def drain(context: Context) -> None:
        host = require_host(context.host, f"{arm.id}-spawn")
        exit_ = await host.close(CLOSE_BUDGET_MS)
        equals(exit_.code, 0, "the fixture host's exit code after stdin EOF")
        # Only the host slot: the NEXT arm's spawn is the authoritative reset.
        context.host = None

    return (
        Segment(
            f"{arm.id}-spawn",
            f"Spawn the canned host playing the {arm.scenario} transcript",
            spawn,
        ),
        Segment(f"{arm.id}-turn", "Start a session and a turn that will need permission", turn),
        Segment(
            f"{arm.id}-request",
            "Take the request off the wire and read the choices the host offers",
            request,
        ),
        Segment(f"{arm.id}-decide", f'Answer with "{arm.choice_id}"', decide),
        Segment(
            f"{arm.id}-resolved",
            "Read the outcome off approval/resolved, then let the turn finish",
            resolved,
        ),
        Segment(f"{arm.id}-drain", "Close stdin and let the fixture host exit cleanly", drain),
    )


SEGMENTS: tuple[Segment[Context], ...] = (*_arm_segments(ALLOW), *_arm_segments(DENY))


async def _run(hosts: RecipeHosts) -> JourneyReport:
    if hosts.conformance_bin is None:
        raise ValueError("conformance_bin is required")
    context = Context(
        conformance_bin=hosts.conformance_bin, transcript_root=hosts.transcript_root
    )

    async def teardown(owned: Context) -> None:
        if owned.host is not None:
            await owned.host.abandon(CLOSE_BUDGET_MS)

    return await run_journey(SEGMENTS, context, teardown)


RECIPE = Recipe(
    id="approve-or-deny",
    title="Approve or deny the agent's permission request",
    docs_page="developer-" "docs/src/content/docs/cookbook/approve-or-deny-a-tool-call.mdx",
    needs=("conformance_bin",),
    run=_run,
)
